x/clp/ante.go

```diff
@@ -0,0 +1,89 @@
+package clp
+
+import (
+	"fmt"
+	"github.com/Sifchain/sifnode/x/clp/keeper"
+	clpTypes "github.com/Sifchain/sifnode/x/clp/types"
+	"github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/pkg/errors"
+)
+
+type SwapFeeChangeDecorator struct {
+	ck keeper.Keeper
+}
+
+func NewSwapFeeChangeDecorator(ck keeper.Keeper) SwapFeeChangeDecorator {
+	return SwapFeeChangeDecorator{
+		ck: ck,
+	}
+}
+
+var _ types.AnteDecorator = SwapFeeChangeDecorator{}
+
+func (r SwapFeeChangeDecorator) AnteHandle(ctx types.Context, tx types.Tx, simulate bool, next types.AnteHandler) (newCtx types.Context, err error) {
+	feeTx, ok := tx.(ante.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+	msg := feeTx.GetMsgs()[0]
+	switch msg := msg.(type) {
+	case clpTypes.MsgSwap:
+		payer := feeTx.FeePayer()
+		// Fee Payer is always the msg signer for swap transactions
+		if !payer.Equals(msg.Signer) {
+			return types.Context{}, errors.New("Fee Payer and MSG Signer are not the same ")
+		}
+		// User is sending rowan , that means they already have rowan balance
+		if msg.SentAsset.Equals(GetSettlementAsset()) {
+			return next(ctx, tx, simulate)
+		}
+		// The amount of Fee user agreed to pay
+		feeInRowan := feeTx.GetFee()
+		requiredRowan := feeInRowan.AmountOf(clpTypes.GetSettlementAsset().Symbol)
```

Comment:
You call `clpTypes.GetSettlementAsset().Symbol` twice. May I suggest storing it in a variable?